docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+|----|---------------|
+|(immediate) child device|`tedge/child/<child-id>/...`|
+|nested child device|`tedge/descendent/<child-id>/...`|
+
+
+.. so that automatic registration can be done for everything coming from `tedge/child/...` topics.
+If keeping that information in the topics is not desired, it can be kept in the payload as well,
+with the caveat that defining static routing rules would not be possible with it.
+
+**Pros**
+
+1. The context on whether the data is coming from the parent, a child or a service is clear from the topics.
+1. Automatic registration is possible at least for services and even child devices with some tweaks.
+
+**Cons**
+
+1. The topics are fairly long and with extensions might easily cross the 7 sub-topic limit of AWS.
+
+## Extended Proposal 1: Include parent id also in the topics
+
+Include the parent `id` information also in the topic as follows:
+
+```
+tedge/<parent-device-id>/<device-id>/<target-type>/...
+```
+
+Examples:
+
+|Target|Topic Prefix|
+|------|-------|
+|tedge device|`tedge/main/device/`|
+|service on tedge device|`tedge/main/service/<service-id>`|
+|immediate child device|`tedge/main/<child-id>/device/`|
+|descendent child device|`tedge/<parent-child-id>/<child-id>/`|
+|service on child device|`tedge/<child-id>/service/<service-id>/`|
+
+For the main device, there is no change as it is the root device,
+unless we want to make it more explicit and symmetric with a `root` prefix like `tedge/root/main/device/`
+
+**Pros**
+
+* When the parent device wants to subscribe to all data from its immediate child devices,
+  it can be done with a `tedge/<parent-child-id>/+/#` subscription.
+  This single subscription will work even when child devices are added dynamically.
+  Otherwise, the parent will have to subscribe for all child devices individually
+  and dynamically subscribe to new child topics, when new child devices are added.
+
+**Cons**
+
+* Topics are even longer for marginal gains.
+* Higher payload size, as the parent name is repeated in each and every message.
+* To subscribe for all nested child devices data as well, the parent will have to do dynamic subscriptions anyway
+  when a new nested child device is added.
+  For example, if a `parent` is already subscribed to `tedge/parent/+/#` for all child devices: `child-1` and `child-2`,
+  when a `child-3` is added, it will be covered with the existing subscription itself.
+  But, if these child devices have nested child devices, the parent will have to subscribe to,
+  `tedge/child-1/+/#` and `tedge/child-2/+/#` separately,
+  and when `child-3` is added, dynamically subscribe to `tedge/child-3/+/#` as well.
+
+## Proposal 2: Unified topics for every "thing"
+
+This proposal is built on top of the following assumptions:
+1. Every entity, a child device, descendent child devices or services, must register with thin-edge
+   to get their unique IDs generated, unless these entities can ensure uniqueness across the entire fleet.
```

Comment:
An implicit registration rule could be added: by default an unregistered thing starting to emit telemetry data is a child-device, not a service, of the main idea. Not sure auto-registration is a good idea though.

Reply:
Sure. But I'm afraid that auto-registration for `services` as well would be highly desired, and that case, we'd need to some parameter to distinguish.

Reply:
Limited auto registration will be supported, but only for `te/device/+/service/+` topics. And the feature can be turned off.